source/http_client_connection.c

```diff
@@ -0,0 +1,551 @@
+
+    py_connection->allocator = allocator;
+
+    struct aws_http_client_connection_options options;
+    AWS_ZERO_STRUCT(options);
+    options.self_size = sizeof(options);
+    options.bootstrap = bootstrap;
+    options.tls_options = connection_options;
+    options.allocator = allocator;
+    options.user_data = py_connection;
+    options.host_name = aws_byte_cursor_from_array((const uint8_t *)host_name, host_name_len);
+    options.port = port_number;
+    options.initial_window_size = initial_window_size;
+    options.socket_options = &socket_options;
+    options.on_setup = s_on_client_connection_setup;
+    options.on_shutdown = s_on_client_connection_shutdown;
+
+    if (aws_http_client_connect(&options)) {
+        PyErr_SetAwsLastError();
+        goto error;
+    }
+
+    Py_RETURN_NONE;
+
+error:
+    if (py_connection) {
+        aws_mem_release(allocator, py_connection);
+    }
+
+    Py_RETURN_NONE;
+}
+
+PyObject *aws_py_http_client_connection_close(PyObject *self, PyObject *args) {
+    (void)self;
+    PyObject *http_impl = NULL;
+
+    if (PyArg_ParseTuple(args, "O", &http_impl)) {
+        if (http_impl) {
+            struct py_http_connection *http_connection =
+                PyCapsule_GetPointer(http_impl, s_capsule_name_http_client_connection);
+            assert(http_connection);
+
+            if (http_connection->connection) {
+                aws_http_connection_close(http_connection->connection);
+            }
+        }
+    }
+
+    Py_RETURN_NONE;
+}
+
+PyObject *aws_py_http_client_connection_is_open(PyObject *self, PyObject *args) {
+    (void)self;
+
+    PyObject *http_impl = NULL;
+
+    if (PyArg_ParseTuple(args, "O", &http_impl)) {
+        if (http_impl) {
+            struct py_http_connection *http_connection =
+                PyCapsule_GetPointer(http_impl, s_capsule_name_http_client_connection);
+            assert(http_connection);
+
+            if (http_connection->connection && aws_http_connection_is_open(http_connection->connection)) {
+                Py_RETURN_TRUE;
+            }
+        }
+    }
+
+    Py_RETURN_FALSE;
+}
+
+struct py_http_stream {
+    struct aws_allocator *allocator;
+    struct aws_http_stream *stream;
+    PyObject *capsule;
+    PyObject *on_stream_completed;
+    PyObject *on_incoming_headers_received;
+    PyObject *on_read_body;
+    PyObject *on_incoming_body;
+    PyObject *received_headers;
+};
+
+static enum aws_http_outgoing_body_state s_stream_outgoing_body(
+    struct aws_http_stream *internal_stream,
+    struct aws_byte_buf *buf,
+    void *user_data) {
+    (void)internal_stream;
+
+    struct py_http_stream *stream = user_data;
+
+    PyGILState_STATE state = PyGILState_Ensure();
+
+    Py_ssize_t data_len = (Py_ssize_t)buf->capacity - buf->len;
+
+    PyObject *mv = PyMemoryView_FromMemory((char *)(buf->buffer + buf->len), data_len, PyBUF_WRITE);
+    PyObject *result = PyObject_CallFunction(stream->on_read_body, "(O)", mv);
```

Comment:
So this version is going to require the callback to be called n+1 times, which is unlike how we handle it elsewhere. Can we not do it the same way for python?

Reply:
not unless we have python return a tuple

Reply:
I'll play with it in a bit and see what I can do